### DISCUSSION TITLE
CouchDB3 constantly at 100% CPU Usage

### DISCUSSION THREAD

**Discussion by skeyby** [apache/couchdb#3019] [Q&A]
Hello,

I have a machine that keeps constantly at 100%. It used to do it with CouchDB 2.1 and now, after an upgrade to CouchDB 3.1, it keeps doing it.

The Tasklist show practically nothing running and even placing the log in debug mode reports nothing interesting.

The machine is mostly idle and just have a local replication job running (from a local db to another local one).

How can I understand what's going on?

I'm attaching the output for "top" from the machine, with thread details open... any suggestion's welcome.

![Schermata 2020-07-21 alle 17 49 17](https://user-images.githubusercontent.com/1609532/88076611-95544d00-cb7a-11ea-92b6-607b5bb3a568.png)

**Comment by SinanGabel** [comment]
You could check your log contents and the size of the logs (maybe also log-rotate if used).

**Reply by skeyby** [reply]
The log's fine the problem is not there...

**Comment by ngspinu** [comment]
@skeyby I went through the same issue a while back. It looks like couchdb is trying to keep the indexes 'warm'. So it, for whatever reasons, keeps indexing non-stop. Check [here](https://docs.couchdb.org/en/master/config/indexbuilds.html?highlight=warm#background-indexing). You could try to reduce the batch_channels, but I have personally set all the _design/docs - to autoupdate: false. If there is any better fix to that I'm looking forward to it. Until then I will hold any other production update to 3.*

**Comment by janl** [comment]
@skeyby do you have any .compact files around that were created with 2.1? If yes, this is it: https://github.com/apache/couchdb/pull/3001

You can safely delete the .compact files (don’t miss the ones for view indexes), and things should compact normally and then things should slow down.

**Comment by ngspinu** [comment]
This still is a big issue. Deleting .compact files - does not seem to fix it. Any other suggestion?
Thank you.

**Comment by ermouth** [comment]
Mb something from https://docs.couchdb.org/en/stable/maintenance/performance.html could help?

**Reply by ngspinu** [reply]
Thanks @ermouth but no. Sadly I've had no luck fixing this by reading the docs.. Thing is - this happens on two different platforms. Cent-os 7 and Mac-Os Catalina.

**Reply by wohali** [reply]
See comment I left over there:

> Check your thresholds for smoosh.
> 
> https://docs.couchdb.org/en/latest/maintenance/compaction.html#channels
> 
> Probably it's the default config for the slack channel(s), which enqueues a db or view for compaction if the difference between actual and file size is > 16MB. This is too small for larger databases. The ratio channel uses a value of 2.0 and this is usually fine on its own.
> 
> Try increasing the values by a few factors of two.

**Comment by Youhan** [comment]
In our case, it was bad coding practice. There was a while loop in our code was constantly running `.changes`

```python
while True:
        changes = db.changes(since=since, include_docs=True, heartbeat=60000)
       # do stuff with changes (if any)...
```

We ended up adding sleep in the while for a quick patch, I don't know if there is a better way to fix it. 
```python
while True:
        time.sleep(sleep_time) # <- added
        changes = db.changes(since=since, include_docs=True, heartbeat=60000)
        # do stuff with changes (if any)...
```
